<div style="background-color: #D18E11; color: white; font-variant: small-caps; padding: 10px; text-align: center;">
  <span style="font-size: 14pt;">Part 01:</span>
  <span style="font-size: 26pt; font-weight: bold;"> Data Engineering</span>
</div>

<div style="background-color: #D18E11; color: white; padding: 10px; text-align: center;">
  <span style="font-size: 14pt; font-weight: bold; font-color: #87B5D6"> Preparing and Structuring U.S. Border Entry Data for Analysis </span><br>  
</div>

<div style="background-color: #D6B276; padding: 10px; margin-top: 10px;">
  <b>Importing Python libraries.</b>
</div>

In [47]:

# Importing required libraries
import pandas as pd
import numpy as np
from IPython.display import display, HTML
import io
import sys
from pandas import DataFrame

<div style="background-color: #D6B276;  padding: 10px;">
<h3>Data Collection</h3> 
</div>

In [48]:

file_path = '../Capstone Project/Border_Crossing_Entry_Data.csv'
df = pd.read_csv(file_path)
raw_df = pd.read_csv(file_path)

# Show the first 5 rows
df.head()

Port Name      State  Port Code            Border      Date  \
0       Jackman      Maine        104  US-Canada Border  Jan 2024   
1      Porthill      Idaho       3308  US-Canada Border  Apr 2024   
2      San Luis    Arizona       2608  US-Mexico Border  Apr 2024   
3  Willow Creek    Montana       3325  US-Canada Border  Jan 2024   
4       Warroad  Minnesota       3423  US-Canada Border  Jan 2024   

                       Measure  Value  Latitude  Longitude  \
0                       Trucks   6556    45.806    -70.397   
1                       Trucks     98    49.000   -116.499   
2                        Buses     10    32.485   -114.782   
3                  Pedestrians      2    49.000   -109.731   
4  Personal Vehicle Passengers   9266    48.999    -95.377   

                           Point  
0   POINT (-70.396722 45.805661)  
1   POINT (-116.49925 48.999861)  
2    POINT (-114.7822222 32.485)  
3  POINT (-109.731333 48.999972)  
4      POINT (-95.376555 48.999)

In [49]:

print("Dataset shape: (rows x columns)", df.shape)

Dataset shape: (rows x columns) (401566, 10)


<div style="background-color: #D6B276; padding: 10px;">
<h3>Data Cleaning</h3> <br>
</div>
<div style="background-color: #D6B276; padding: 10px;">
Checking for the null or missing values in dataset and the data type of the columns.
</div>

In [50]:

# Capture df.info() output as text
buffer = io.StringIO()
df.info(buf=buffer)
info_output = buffer.getvalue()

# Get null summary
null_summary = df.isnull().sum().to_frame(name='Null Count')

# Display side-by-side using HTML
display(HTML(f"""
<div style="display: flex; justify-content: space-between;">
  <div style="flex: 1; margin-right: 20px;">
    <h4>DataFrame Info</h4>
    <pre>{info_output}</pre>
  </div>
  <div style="flex: 1;">
    <h4>Null Count per Column</h4>
    {null_summary.to_html()}
  </div>
</div>
"""))


<div style="background-color: #D6B276; padding: 10px;">
The above output shows that there are no null values in the dataset. <br>
And the datatype of the <b>Date</b> column is object which needs to be in data-time format. <br>
Converting the Date column into date-time format.
</div>

In [51]:

df['Date'] = pd.to_datetime(df['Date'], format='%b %Y')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 401566 entries, 0 to 401565
Data columns (total 10 columns):
 #   Column     Non-Null Count   Dtype         
---  ------     --------------   -----         
 0   Port Name  401566 non-null  object        
 1   State      401566 non-null  object        
 2   Port Code  401566 non-null  int64         
 3   Border     401566 non-null  object        
 4   Date       401566 non-null  datetime64[ns]
 5   Measure    401566 non-null  object        
 6   Value      401566 non-null  int64         
 7   Latitude   401566 non-null  float64       
 8   Longitude  401566 non-null  float64       
 9   Point      401566 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(2), object(5)
memory usage: 30.6+ MB


<div style="background-color: #D6B276; padding: 10px;">Checking for duplicate values in the dataset.
</div>

In [52]:

print("Number of rows before removing duplicate rows:\033[1m", df.shape[0],"\033[0m") 

Number of rows before removing duplicate rows: 401566 


In [53]:

duplicate_rows = df.duplicated().sum()
print("Number of duplicate rows:\033[1m", duplicate_rows,"\033[0m")

Number of duplicate rows: 10 


In [54]:

# Droping duplicate rows
df = df.drop_duplicates()
print("Number of rows after removing duplicate rows:\033[1m", df.shape[0],"\033[0m") 

Number of rows after removing duplicate rows: 401556 


<div style="background-color: #D6B276; padding: 10px;">
<h3>Data Transformation </h3><br>
</div>
<div style="background-color: #D6B276; padding: 10px;">
<b>Date-Time Parsing</b><br>
Adding two new columns 'Month' and 'Year' to analyse the data periodically.
</div>

In [55]:

df['Month'] = df['Date'].dt.month
df['Year'] = df['Date'].dt.year
df.info()
# Previewing the new columns
df[['Date', 'Year', 'Month']].head()

<class 'pandas.core.frame.DataFrame'>
Index: 401556 entries, 0 to 401565
Data columns (total 12 columns):
 #   Column     Non-Null Count   Dtype         
---  ------     --------------   -----         
 0   Port Name  401556 non-null  object        
 1   State      401556 non-null  object        
 2   Port Code  401556 non-null  int64         
 3   Border     401556 non-null  object        
 4   Date       401556 non-null  datetime64[ns]
 5   Measure    401556 non-null  object        
 6   Value      401556 non-null  int64         
 7   Latitude   401556 non-null  float64       
 8   Longitude  401556 non-null  float64       
 9   Point      401556 non-null  object        
 10  Month      401556 non-null  int32         
 11  Year       401556 non-null  int32         
dtypes: datetime64[ns](1), float64(2), int32(2), int64(2), object(5)
memory usage: 36.8+ MB


Date  Year  Month
0 2024-01-01  2024      1
1 2024-04-01  2024      4
2 2024-04-01  2024      4
3 2024-01-01  2024      1
4 2024-01-01  2024      1

<div style="background-color: #D6B276; padding: 10px;">
<b>Encoding Categorial Variable</b><br>
Creating new tables (lookup table) for <b>state</b>, <b>border</b> and <b>measure</b> values and assigning them <b>code</b>.<br><br>
Creating state_table.
</div>

In [56]:

# Extracting unique values from state column
unique_states = df['State'].unique()

# Displaying the unique states one below the other with the count
print(f"Number of unique states:\033[1m {len(unique_states)} \033[0m \n")

# Creating a dictionary to map unique state names to three-digit codes
state_codes = {state: f"{i+1:03}" for i, state in enumerate(unique_states)}

# Display the state and their respective three-digit codes
for state, code in state_codes.items():
    print(f"{state}: {code}")

Number of unique states: 14  

Maine: 001
Idaho: 002
Arizona: 003
Montana: 004
Minnesota: 005
Texas: 006
North Dakota: 007
Washington: 008
Alaska: 009
California: 010
Michigan: 011
New York: 012
New Mexico: 013
Vermont: 014


In [57]:

# Create a new Table with two columns: 'Name' and 'Code'
state_table = pd.DataFrame(list(state_codes.items()), columns=['Name', 'Code'])
print(state_table)

            Name Code
0          Maine  001
1          Idaho  002
2        Arizona  003
3        Montana  004
4      Minnesota  005
5          Texas  006
6   North Dakota  007
7     Washington  008
8         Alaska  009
9     California  010
10      Michigan  011
11      New York  012
12    New Mexico  013
13       Vermont  014


In [58]:

# Find duplicate State Names
duplicates = state_table[state_table.duplicated(subset=['Name'], keep=False)]

if duplicates.empty:
    print("No duplicate state names found.")
else:
    print("Duplicate state names found:")
    print(duplicates)

No duplicate state names found.


<div style="background-color: #D6B276; padding: 10px;">
Creating border_table.
</div>

In [59]:

# Extracting unique values from border column
unique_border = df['Border'].unique()

# Displaying the unique border one below the other with the count
print(f"Number of unique border:\033[1m {len(unique_border)} \033[0m \n")

# Creating a dictionary to map unique border names to three-digit codes
border_codes = {border: f"{i+1:03}" for i, border in enumerate(unique_border)}

# Display the border and their respective three-digit codes
for border, code in border_codes.items():
    print(f"{border}: {code}")

Number of unique border: 2  

US-Canada Border: 001
US-Mexico Border: 002


In [60]:

# Create a new DataFrame with two columns: 'Name' and 'Code'
border_table = pd.DataFrame(list(border_codes.items()), columns=['Name', 'Code'])
print(border_table)

               Name Code
0  US-Canada Border  001
1  US-Mexico Border  002


<div style="background-color: #D6B276; padding: 10px;">
Creating measure_table.
</div>

In [61]:

# Extracting unique values from measure column
unique_measure = df['Measure'].unique()

# Displaying the unique measure one below the other with the count
print(f"Number of unique measure:\033[1m {len(unique_measure)}\033[0m \n")

# Creating a dictionary to map unique measure names to three-digit codes
measure_codes = {measure: f"{i+1:03}" for i, measure in enumerate(unique_measure)}

# Display the measure and their respective three-digit codes
for measure, code in measure_codes.items():
    print(f"{measure}: {code}")

Number of unique measure: 12 

Trucks: 001
Buses: 002
Pedestrians: 003
Personal Vehicle Passengers: 004
Personal Vehicles: 005
Bus Passengers: 006
Truck Containers Loaded: 007
Truck Containers Empty: 008
Rail Containers Empty: 009
Trains: 010
Train Passengers: 011
Rail Containers Loaded: 012


In [62]:

# Create a new DataFrame with two columns: 'Name' and 'Code'
measure_table = pd.DataFrame(list(measure_codes.items()), columns=['Name', 'Code'])
print(measure_table)

                           Name Code
0                        Trucks  001
1                         Buses  002
2                   Pedestrians  003
3   Personal Vehicle Passengers  004
4             Personal Vehicles  005
5                Bus Passengers  006
6       Truck Containers Loaded  007
7        Truck Containers Empty  008
8         Rail Containers Empty  009
9                        Trains  010
10             Train Passengers  011
11       Rail Containers Loaded  012


In [63]:

# Find duplicate Measure Names
duplicates = measure_table[measure_table.duplicated(subset=['Name'], keep=False)]

if duplicates.empty:
    print("No duplicate measure names found.")
else:
    print("Duplicate measure names found:")
    print(duplicates)

No duplicate measure names found.


<div style="background-color: #D6B276; padding: 10px;">
Since, there is port code available in dataset for port there is no need to create a new code. <br> So, extracting unique portname and portcode from the original dataset and creating a new table.
</div>

In [64]:

# Extracting unique values
unique_ports = df[['Port Name', 'Port Code']].drop_duplicates()

# Sort by PortCode
unique_ports_sorted = unique_ports.sort_values(by='Port Code')

#Creating new table for port names and code
port_table = unique_ports.sort_values(by='Port Code')
port_table.head()

Port Name  Port Code
3917   Portland        101
131    Eastport        103
0       Jackman        104
39    Vanceboro        105
153     Houlton        106

In [65]:

# Find duplicate Port Names
duplicates = port_table[port_table.duplicated(subset=['Port Name'], keep=False)]

if duplicates.empty:
    print("No duplicate port names found.")
else:
    print("Duplicate port names found:")
    print(duplicates)

Duplicate port names found:
    Port Name  Port Code
131  Eastport        103
54   Eastport       3302


In [66]:

# Filter rows where Port Name is 'Eastport' (case-sensitive)
filtered_df = df[df['Port Name'] == 'Eastport']

# Select only the columns State, Port Name, and Port Code
result = filtered_df[['State', 'Port Name', 'Port Code']].drop_duplicates()
result

State Port Name  Port Code
54   Idaho  Eastport       3302
131  Maine  Eastport        103

In [67]:

# Replace 'Eastport' with 'Eastport_ID' where Port Code is 3302
port_table.loc[(port_table['Port Name'] == 'Eastport') & (port_table['Port Code'] == 3302), 'Port Name'] = 'Eastport_ID'

# Replace 'Eastport' with 'Eastport_ME' where Port Code is 103
port_table.loc[(port_table['Port Name'] == 'Eastport') & (port_table['Port Code'] == 103), 'Port Name'] = 'Eastport_ME'

# Display updated dataframe
port_table.head()


Port Name  Port Code
3917     Portland        101
131   Eastport_ME        103
0         Jackman        104
39      Vanceboro        105
153       Houlton        106

<div style="background-color: #D6B276; padding: 10px;">
Generating <b>state abbreviations</b> in state_table for visualizations.
</div>

In [68]:

# Dictionary of state abbreviations
state_abbr = {
    "Maine": "ME",
    "Idaho": "ID",
    "Arizona": "AZ",
    "Montana": "MT",
    "Minnesota": "MN",
    "Texas": "TX",
    "North Dakota": "ND",
    "Washington": "WA",
    "Alaska": "AK",
    "California": "CA",
    "Michigan": "MI",
    "New York": "NY",
    "New Mexico": "NM",
    "Vermont": "VT"
}

# Add new column 'Abbr' by mapping 'Name' column
state_table['Abbr'] = state_table['Name'].map(state_abbr)

# Display result
state_table.head(20)

Name Code Abbr
0          Maine  001   ME
1          Idaho  002   ID
2        Arizona  003   AZ
3        Montana  004   MT
4      Minnesota  005   MN
5          Texas  006   TX
6   North Dakota  007   ND
7     Washington  008   WA
8         Alaska  009   AK
9     California  010   CA
10      Michigan  011   MI
11      New York  012   NY
12    New Mexico  013   NM
13       Vermont  014   VT

<div style="background-color: #D6B276; padding: 10px;">
Generating <b>state co-ordinates</b> in state_table for visualizations.
</div>

In [69]:

# Dictionary of state abbreviations and their coordinates
state_coordinates = {
    "ME": {"lat": 44.6937, "lon": -69.3819},
    "ID": {"lat": 44.0682, "lon": -114.7420},
    "AZ": {"lat": 34.0489, "lon": -111.0937},
    "MT": {"lat": 46.8797, "lon": -110.3626},
    "MN": {"lat": 46.7296, "lon": -94.6859},
    "TX": {"lat": 31.9686, "lon": -99.9018},
    "ND": {"lat": 47.5515, "lon": -101.0020},
    "WA": {"lat": 47.7511, "lon": -120.7401},
    "AK": {"lat": 61.3850, "lon": -152.2683},
    "CA": {"lat": 36.7783, "lon": -119.4179},
    "MI": {"lat": 44.3148, "lon": -85.6024},
    "NY": {"lat": 40.7128, "lon": -74.0060},
    "NM": {"lat": 34.5199, "lon": -105.8701},
    "VT": {"lat": 44.5588, "lon": -72.5778}
}

# Add the Latitude and Longitude columns to the state_table using the Abbr as key
state_table['Latitude'] = state_table['Abbr'].map(lambda abbr: state_coordinates[abbr]['lat'])
state_table['Longitude'] = state_table['Abbr'].map(lambda abbr: state_coordinates[abbr]['lon'])

# Display the updated table
state_table.head(20)

Name Code Abbr  Latitude  Longitude
0          Maine  001   ME   44.6937   -69.3819
1          Idaho  002   ID   44.0682  -114.7420
2        Arizona  003   AZ   34.0489  -111.0937
3        Montana  004   MT   46.8797  -110.3626
4      Minnesota  005   MN   46.7296   -94.6859
5          Texas  006   TX   31.9686   -99.9018
6   North Dakota  007   ND   47.5515  -101.0020
7     Washington  008   WA   47.7511  -120.7401
8         Alaska  009   AK   61.3850  -152.2683
9     California  010   CA   36.7783  -119.4179
10      Michigan  011   MI   44.3148   -85.6024
11      New York  012   NY   40.7128   -74.0060
12    New Mexico  013   NM   34.5199  -105.8701
13       Vermont  014   VT   44.5588   -72.5778

<div style="background-color: #D6B276; padding: 10px;">
<b> Creating month_table for month mapping dictionary. <b>
</b>

In [70]:

month_mapping = {
    'January': '01',
    'February': '02',
    'March': '03',
    'April': '04',
    'May': '05',
    'June': '06',
    'July': '07',
    'August': '08',
    'September': '09',
    'October': '10',
    'November': '11',
    'December': '12'
}

month_table = pd.DataFrame(list(month_mapping.items()), columns=['Month', 'Month_Number'])

month_table

Month Month_Number
0     January           01
1    February           02
2       March           03
3       April           04
4         May           05
5        June           06
6        July           07
7      August           08
8   September           09
9     October           10
10   November           11
11   December           12

<div style="background-color: #D6B276; padding: 10px;">
<h3> Feature Engineering </h3><br>
</div>
<div style="background-color: #D6B276; padding: 10px;">
Merging the newly created codes for state, border and measure into the working dataset.<br>
Starting by creating a copy of original dataframe and working on it before deleting.
</div>

In [71]:

# Create a copy of the original DataFrame and name it 'refined_df'
refined_df = df.copy()

# Display the new DataFrame to confirm
refined_df.head()

Port Name      State  Port Code            Border       Date  \
0       Jackman      Maine        104  US-Canada Border 2024-01-01   
1      Porthill      Idaho       3308  US-Canada Border 2024-04-01   
2      San Luis    Arizona       2608  US-Mexico Border 2024-04-01   
3  Willow Creek    Montana       3325  US-Canada Border 2024-01-01   
4       Warroad  Minnesota       3423  US-Canada Border 2024-01-01   

                       Measure  Value  Latitude  Longitude  \
0                       Trucks   6556    45.806    -70.397   
1                       Trucks     98    49.000   -116.499   
2                        Buses     10    32.485   -114.782   
3                  Pedestrians      2    49.000   -109.731   
4  Personal Vehicle Passengers   9266    48.999    -95.377   

                           Point  Month  Year  
0   POINT (-70.396722 45.805661)      1  2024  
1   POINT (-116.49925 48.999861)      4  2024  
2    POINT (-114.7822222 32.485)      4  2024  
3  POINT (-109.731333 48.999972)      1  2024  
4      POINT (-95.376555 48.999)      1  2024

<div style="background-color: #D6B276; padding: 10px;">
Merging State_code with the working dataset.
</div>

In [72]:

# Mapping state_table has the columns 'Name' and 'Code'
state_table.rename(columns={'Name': 'State', 'Code': 'State_Code'}, inplace=True)

# Merge the refined_df with state_table based on 'State' column
refined_df = pd.merge(refined_df, state_table[['State', 'State_Code']], on='State', how='left')

# Display the updated refined_df
refined_df.head(2)

Port Name  State  Port Code            Border       Date Measure  Value  \
0   Jackman  Maine        104  US-Canada Border 2024-01-01  Trucks   6556   
1  Porthill  Idaho       3308  US-Canada Border 2024-04-01  Trucks     98   

   Latitude  Longitude                         Point  Month  Year State_Code  
0    45.806    -70.397  POINT (-70.396722 45.805661)      1  2024        001  
1    49.000   -116.499  POINT (-116.49925 48.999861)      4  2024        002

<div style="background-color: #D6B276; padding: 10px;">
Merging border_code with the working dataset.
</div>

In [73]:

# Assuming border_code has the columns 'Name' and 'Code'
border_table.rename(columns={'Name': 'Border', 'Code': 'Border_Code'}, inplace=True)

# Merge the refined_df with border_table based on 'Border' column
refined_df = pd.merge(refined_df, border_table[['Border', 'Border_Code']], on='Border', how='left')

# Display the updated refined_df
refined_df.head(2)

Port Name  State  Port Code            Border       Date Measure  Value  \
0   Jackman  Maine        104  US-Canada Border 2024-01-01  Trucks   6556   
1  Porthill  Idaho       3308  US-Canada Border 2024-04-01  Trucks     98   

   Latitude  Longitude                         Point  Month  Year State_Code  \
0    45.806    -70.397  POINT (-70.396722 45.805661)      1  2024        001   
1    49.000   -116.499  POINT (-116.49925 48.999861)      4  2024        002   

  Border_Code  
0         001  
1         001

<div style="background-color: #D6B276; padding: 10px;">
Merging measure_code with the working dataset.
</div>

In [74]:

# Assuming measure_code has the columns 'Name' and 'Code'
measure_table.rename(columns={'Name': 'Measure', 'Code': 'Measure_Code'}, inplace=True)

# Merge the refined_df with measure_table based on 'Measure' column
refined_df = pd.merge(refined_df, measure_table[['Measure', 'Measure_Code']], on='Measure', how='left')

# Display the updated refined_df
refined_df.head(2)

Port Name  State  Port Code            Border       Date Measure  Value  \
0   Jackman  Maine        104  US-Canada Border 2024-01-01  Trucks   6556   
1  Porthill  Idaho       3308  US-Canada Border 2024-04-01  Trucks     98   

   Latitude  Longitude                         Point  Month  Year State_Code  \
0    45.806    -70.397  POINT (-70.396722 45.805661)      1  2024        001   
1    49.000   -116.499  POINT (-116.49925 48.999861)      4  2024        002   

  Border_Code Measure_Code  
0         001          001  
1         001          001

<div style="background-color: #D6B276; padding: 10px;">
Verifying the correctness of merged columns.
</div>

In [75]:

# comparing the values for the state in both tables
comparison_df = pd.merge(refined_df[['State', 'State_Code']], state_table[['State', 'State_Code']], on='State', how='left')

# Rename columns to include table names for better understanding
comparison_df.rename(columns={
    'State': 'State_from_refined_df',
    'State_Code_x': 'State_Code_from_refined_df',  # 'State_Code_x' is from refined_df
    'State_Code_y': 'State_Code_from_state_table'  # 'State_Code_y' is from state_table
}, inplace=True)

# Remove duplicates and show unique rows only
unique_comparison_df = comparison_df.drop_duplicates()

# Display the comparison table
unique_comparison_df

State_from_refined_df State_Code_from_refined_df  \
0                  Maine                        001   
1                  Idaho                        002   
2                Arizona                        003   
3                Montana                        004   
4              Minnesota                        005   
6                  Texas                        006   
11          North Dakota                        007   
16            Washington                        008   
18                Alaska                        009   
19            California                        010   
25              Michigan                        011   
30              New York                        012   
37            New Mexico                        013   
75               Vermont                        014   

   State_Code_from_state_table  
0                          001  
1                          002  
2                          003  
3                          004  
4                          005  
6                          006  
11                         007  
16                         008  
18                         009  
19                         010  
25                         011  
30                         012  
37                         013  
75                         014

In [76]:

# comparing the values for the border in both tables
comparison_df = pd.merge(refined_df[['Border', 'Border_Code']], border_table[['Border', 'Border_Code']], on='Border', how='left')

# Rename columns to include table names for better understanding
comparison_df.rename(columns={
    'Border': 'Border_from_refined_df',
    'Border_Code_x': 'Border_Code_from_refined_df',  # 'Border_Code_x' is from refined_df
    'Border_Code_y': 'Border_Code_from_state_table'  # 'Border_Code_y' is from border_table
}, inplace=True)

# Remove duplicates and show unique rows only
unique_comparison_df = comparison_df.drop_duplicates()

# Display the comparison table
unique_comparison_df

Border_from_refined_df Border_Code_from_refined_df  \
0       US-Canada Border                         001   
2       US-Mexico Border                         002   

  Border_Code_from_state_table  
0                          001  
2                          002

In [77]:

# comparing the values for the measure in both tables
comparison_df = pd.merge(refined_df[['Measure', 'Measure_Code']], measure_table[['Measure', 'Measure_Code']], on='Measure', how='left')

# Rename columns to include table names for better understanding
comparison_df.rename(columns={
    'Measure': 'State_from_refined_df',
    'Measure_Code_x': 'Measure_Code_from_refined_df',  # 'Measure_Code_x' is from refined_df
    'Measure_Code_y': 'Measure_Code_from_state_table'  # 'Measure_Code_y' is from measure_table
}, inplace=True)

# Remove duplicates and show unique rows only
unique_comparison_df = comparison_df.drop_duplicates()

# Display the comparison table
unique_comparison_df

State_from_refined_df Measure_Code_from_refined_df  \
0                        Trucks                          001   
2                         Buses                          002   
3                   Pedestrians                          003   
4   Personal Vehicle Passengers                          004   
5             Personal Vehicles                          005   
10               Bus Passengers                          006   
11      Truck Containers Loaded                          007   
13       Truck Containers Empty                          008   
15        Rail Containers Empty                          009   
21                       Trains                          010   
26             Train Passengers                          011   
45       Rail Containers Loaded                          012   

   Measure_Code_from_state_table  
0                            001  
2                            002  
3                            003  
4                            004  
5                            005  
10                           006  
11                           007  
13                           008  
15                           009  
21                           010  
26                           011  
45                           012

<div style="background-color: #D6B276; padding: 10px;">
Removing the <i>border</i> from the values of 'Name' column from border_table.
</div>

In [78]:

border_table['Border'] = border_table['Border'].replace({
    'US-Canada Border': 'US-Canada',
    'US-Mexico Border': 'US-Mexico'
})
print(border_table)

      Border Border_Code
0  US-Canada         001
1  US-Mexico         002


<div style="background-color: #D6B276; padding: 10px;">
As the codes for <i>'State'</i>, <i>'Border'</i> and <i>'Measure'</i> columns have already merged into the working dataset, these column should be removed to avoid data redundancy.<br>
The code for <i>'Port Name'</i> column was originally provided in the raw dataset.<br>
Finally, the <i>'Point'</i> column conatins location coordinates that are already separated into <i>'Latitude'</i> and <i>'Longitude'</i> columns in raw dataset.<br>
Removing these columns helps reduce noise and dimensionality in the working dataset.
</div>

In [79]:

refined_df = refined_df.drop(columns=["Port Name", "State", "Border", "Measure", "Point"])
refined_df.head(2)

Port Code       Date  Value  Latitude  Longitude  Month  Year State_Code  \
0        104 2024-01-01   6556    45.806    -70.397      1  2024        001   
1       3308 2024-04-01     98    49.000   -116.499      4  2024        002   

  Border_Code Measure_Code  
0         001          001  
1         001          001

<div style="background-color: #D6B276; padding: 10px;">
Comapring the memory after doing feature engineering by creating code and removing redundant data.
</div>

In [80]:

# List the names of the DataFrames to fetch the size
dataframe_names = ['raw_df', 'refined_df', 'state_table', 'border_table','measure_table','port_table','month_table']

for name in dataframe_names:
    df = globals().get(name)
    if isinstance(df, DataFrame):
        mem = df.memory_usage(deep=True).sum()  # in bytes
        mem_MB = mem / (1024 ** 2)  # convert to MB
        print(f"{name}: {mem_MB:.2f} MB")
    else:
        print(f"{name} is not a DataFrame or does not exist.")

raw_df: 157.62 MB
refined_df: 78.12 MB
state_table: 0.00 MB
border_table: 0.00 MB
measure_table: 0.00 MB
port_table: 0.01 MB
month_table: 0.00 MB


<div style="background-color: #D6B276; padding: 10px;">
Following <b>data engineering</b>, the memory consumption of the raw source was <b>decreased from 157.62 MB to 78.12 MB</b>, representing almost a <b>50% reduction</b>.<br><br>
Next, dentifying the dataframes that we created during the process up to this point and deleting the unwanted ones.<br>
</div>

In [81]:

# Get all DataFrames excluding system/temp vars starting with underscore
dataframes = {
    name: obj for name, obj in globals().items() 
    if isinstance(obj, DataFrame) and not name.startswith('_')
}

exclude_names = {'df'}
filtered_dfs = {name: df for name, df in dataframes.items() if name not in exclude_names}

print("User DataFrames:")
for name in filtered_dfs:
    print(name)

User DataFrames:
state_table
duplicates
border_table
measure_table
port_table
filtered_df
result
month_table
finalized_broder_crossing_df
dataframe_dict
column_dict
raw_df
null_summary
unique_ports
unique_ports_sorted
refined_df
comparison_df
unique_comparison_df


<div style="background-color: #D6B276; padding: 10px;">
<b>How DataFrames affects performance in Python Environment</b><br>
1. Memory Usage<br>
   Large DataFrames consume more RAM. If you run out of memory, your system may slow down or crash.<br>
   Inefficient data types (like using object for strings instead of category or numeric types) can bloat memory usage.<br>
2. Computation Speed<br>
   Operations on bigger DataFrames naturally take longer.<br>
   Complex transformations or joins can be slow if the DataFrame is huge.<br>
3. Garbage Collection/Memory Management<br>
   Keeping many DataFrames around, especially large ones, means Python’s garbage collector has more work, possibly causing pauses.<br>
   Not deleting or clearing unused DataFrames leads to wasted memory.<br>
4. I/O Performance<br>
   Reading/writing large DataFrames from disk or network can slow down your workflow.<br>
</div>

In [82]:

#deleting unmanted dataframes
del null_summary
del unique_ports
del unique_ports_sorted
del comparison_df
del raw_df
del unique_comparison_df

<div style="background-color: #D6B276; padding: 10px;">
Retrieving the minimumn and maximun dates to filter the dataset for project use.
</div>

In [83]:

min_year = refined_df['Date'].min()
max_year = refined_df['Date'].max()

print(f"Minimum year: \033[1m {min_year} \033[0m ")
print(f"Maximum year: \033[1m {max_year} \033[0m ")

Minimum year:  1996-01-01 00:00:00  
Maximum year:  2025-05-01 00:00:00  


<div style="background-color: #D6B276; padding: 10px;">
This project analyzes data from the <b>past decade (2015-2024)</b>.
</div>

In [84]:

num_rows = refined_df.shape[0]
print(f"Number of rows before trimming: \033[1m {num_rows} \033[0m")

Number of rows before trimming:  401556 


In [85]:

#Filtering the datset for the year 2019 to 2024
refined_df = refined_df[(refined_df['Year'] >= 2015) & (refined_df['Year'] <= 2024)]

In [86]:

min_year = refined_df['Date'].min()
max_year = refined_df['Date'].max()

print(f"Minimum year: \033[1m {min_year} \033[0m")
print(f"Maximum year: \033[1m {max_year} \033[0m")

Minimum year:  2015-01-01 00:00:00 
Maximum year:  2024-12-01 00:00:00 


In [87]:

num_rows = refined_df.shape[0]
print(f"Number of rows after trimming: \033[1m {num_rows} \033[0m")

Number of rows after trimming:  102519 


<div style="background-color: #D6B276; padding: 10px;">
Renaming the <i>refined_df</i> to <i>finalized_broder_crossing_df</i>.
</div>

In [88]:

finalized_broder_crossing_df = refined_df 
del refined_df   
finalized_broder_crossing_df.head(2)

Port Code       Date  Value  Latitude  Longitude  Month  Year State_Code  \
0        104 2024-01-01   6556    45.806    -70.397      1  2024        001   
1       3308 2024-04-01     98    49.000   -116.499      4  2024        002   

  Border_Code Measure_Code  
0         001          001  
1         001          001

<div style="background-color: #D6B276; padding: 10px;">
<h3> Data Dictionary </h3><br>
</div>
<div style="background-color: #D6B276; padding: 10px;">
<b>Importance of data dictionary.</b><br>
Improves understanding and collaboration.<br>
Aids in documentation and reproductibility.<br>
Facilitates data governance.<br>
Supports data cleaning and validation.<br>
Sppeds up analysis and development.<br>
</div>

In [89]:

# Put your DataFrames in a dictionary with names as keys
dfs = {
    'finalized_broder_crossing_df': finalized_broder_crossing_df,
    'state_table': state_table,
    'border_table': border_table,
    'measure_table': measure_table,
    'port_table': port_table,
    'month_table': month_table
}

# Generate DataFrame codes like dfc_001, dfc_002, ...
def generate_df_codes(dfs):
    codes = {}
    for i, name in enumerate(dfs.keys(), 1):
        codes[name] = f'dfc_{i:03d}'
    return codes

df_codes = generate_df_codes(dfs)

# DataFrame-level dictionary with code column
def get_dataframe_dict(dfs: dict, df_codes: dict) -> pd.DataFrame:
    records = []
    for name, df in dfs.items():
        records.append({
            'DataFrame Code': df_codes[name],
            'DataFrame Name': name,
            'Description': '',
            'Num Rows': df.shape[0],
            'Num Columns': df.shape[1]
        })
    return pd.DataFrame(records)

# Column-level dictionary using DataFrame code instead of name
def get_column_dict(dfs: dict, df_codes: dict) -> pd.DataFrame:
    records = []
    for df_name, df in dfs.items():
        for col in df.columns:
            records.append({
                'DataFrame Code': df_codes[df_name],
                'Column Name': col,
                'Data Type': df[col].dtype,
                'Missing Values': df[col].isnull().sum(),
                'Unique Values': df[col].nunique(),
                'Example Value': df[col].dropna().iloc[0] if df[col].dropna().size > 0 else None,
                'Description': ''
            })
    return pd.DataFrame(records)

# Create dictionaries
dataframe_dict = get_dataframe_dict(dfs, df_codes)
column_dict = get_column_dict(dfs, df_codes)

# Display results
print("DataFrame Dictionary:")
display(dataframe_dict)

print("\nColumn Dictionary:")
display(column_dict)

DataFrame Dictionary:


DataFrame Code                DataFrame Name Description  Num Rows  \
0        dfc_001  finalized_broder_crossing_df                102519   
1        dfc_002                   state_table                    14   
2        dfc_003                  border_table                     2   
3        dfc_004                 measure_table                    12   
4        dfc_005                    port_table                   117   
5        dfc_006                   month_table                    12   

   Num Columns  
0           10  
1            5  
2            2  
3            2  
4            2  
5            2


Column Dictionary:


DataFrame Code   Column Name       Data Type  Missing Values  \
0         dfc_001     Port Code           int64               0   
1         dfc_001          Date  datetime64[ns]               0   
2         dfc_001         Value           int64               0   
3         dfc_001      Latitude         float64               0   
4         dfc_001     Longitude         float64               0   
5         dfc_001         Month           int32               0   
6         dfc_001          Year           int32               0   
7         dfc_001    State_Code          object               0   
8         dfc_001   Border_Code          object               0   
9         dfc_001  Measure_Code          object               0   
10        dfc_002         State          object               0   
11        dfc_002    State_Code          object               0   
12        dfc_002          Abbr          object               0   
13        dfc_002      Latitude         float64               0   
14        dfc_002     Longitude         float64               0   
15        dfc_003        Border          object               0   
16        dfc_003   Border_Code          object               0   
17        dfc_004       Measure          object               0   
18        dfc_004  Measure_Code          object               0   
19        dfc_005     Port Name          object               0   
20        dfc_005     Port Code           int64               0   
21        dfc_006         Month          object               0   
22        dfc_006  Month_Number          object               0   

    Unique Values        Example Value Description  
0             115                  104              
1             120  2024-01-01 00:00:00              
2           27874                 6556              
3              71               45.806              
4             114              -70.397              
5              12                    1              
6              10                 2024              
7              14                  001              
8               2                  001              
9              12                  001              
10             14                Maine              
11             14                  001              
12             14                   ME              
13             14              44.6937              
14             14             -69.3819              
15              2            US-Canada              
16              2                  001              
17             12               Trucks              
18             12                  001              
19            117             Portland              
20            117                  101              
21             12              January              
22             12                   01

In [90]:

# Dataframe codes: dfc_001, dfc_002, dfc_003, dfc_004, dfc_005, dfc_006
display(
    pd.merge(column_dict, dataframe_dict, on='DataFrame Code', how='left')
      .query("`DataFrame Code` == 'dfc_001'")
      [[
          'DataFrame Name',
          'Column Name',
          'Data Type',
          'Missing Values',
          'Unique Values',
          'Example Value',
      ]]
)

DataFrame Name   Column Name       Data Type  Missing Values  \
0  finalized_broder_crossing_df     Port Code           int64               0   
1  finalized_broder_crossing_df          Date  datetime64[ns]               0   
2  finalized_broder_crossing_df         Value           int64               0   
3  finalized_broder_crossing_df      Latitude         float64               0   
4  finalized_broder_crossing_df     Longitude         float64               0   
5  finalized_broder_crossing_df         Month           int32               0   
6  finalized_broder_crossing_df          Year           int32               0   
7  finalized_broder_crossing_df    State_Code          object               0   
8  finalized_broder_crossing_df   Border_Code          object               0   
9  finalized_broder_crossing_df  Measure_Code          object               0   

   Unique Values        Example Value  
0            115                  104  
1            120  2024-01-01 00:00:00  
2          27874                 6556  
3             71               45.806  
4            114              -70.397  
5             12                    1  
6             10                 2024  
7             14                  001  
8              2                  001  
9             12                  001

In [91]:

# Get all DataFrames excluding system/temp vars starting with underscore
dataframes = {
    name: obj for name, obj in globals().items() 
    if isinstance(obj, DataFrame) and not name.startswith('_')
}

exclude_names = {'df'}
filtered_dfs = {name: df for name, df in dataframes.items() if name not in exclude_names}

print("User DataFrames:")
for name in filtered_dfs:
    print(name)

User DataFrames:
state_table
duplicates
border_table
measure_table
port_table
filtered_df
result
month_table
finalized_broder_crossing_df
dataframe_dict
column_dict


In [92]:

# Exporting data to CSV file
state_table.to_csv('state_table.csv', index=False)
border_table.to_csv('border_table.csv', index=False)
measure_table.to_csv('measure_table.csv', index=False)
port_table.to_csv('port_table.csv', index=False)
month_table.to_csv('month_table.csv', index=False)
finalized_broder_crossing_df.to_csv('finalized_broder_crossing_df.csv', index=False)
dataframe_dict.to_csv('dataframe_dict.csv', index=False)
column_dict.to_csv('column_dict.csv', index=False)
print('CSV files are created for: \nstate_table\nborder_table\nmeasure_table\nport_table\nmonth_table\nfinalized_broder_crossing_df\ndataframe_dict\ncolumn_dict')

CSV files are created for: 
state_table
border_table
measure_table
port_table
month_table
finalized_broder_crossing_df
dataframe_dict
column_dict
